qutebrowser/browser/webengine/notification.py

```diff
@@ -0,0 +1,134 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Handles sending notifications over DBus."""
+
+from typing import Dict, Any
+
+from qutebrowser.utils import log
+
+from PyQt5.QtGui import QImage
+from PyQt5.QtCore import QVariant, QMetaType, QByteArray
+from PyQt5.QtDBus import QDBusConnection, QDBusInterface, QDBus, QDBusArgument
+from PyQt5.QtWebEngineCore import QWebEngineNotification
+from PyQt5.QtWebEngineWidgets import QWebEngineProfile
+
+
+class DBusException(Exception):
+    """Raised when something goes wrong with talking to DBus."""
+
+
+class DBusNotificationManager:
+    """Manages notifications that are sent over DBus."""
+
+    def __init__(self):
+        # Dict mapping notification IDs to the corresponding QWebEngine
+        # notification objects.
+        self.active_notifications = {}
+        bus = QDBusConnection.sessionBus()
+        if not bus.isConnected():
+            raise DBusException("Failed to connect to DBus session bus")
+
+        self.interface = QDBusInterface(
+            "org.freedesktop.Notifications",  # service
+            "/org/freedesktop/Notifications",  # path
+            "org.freedesktop.Notifications",  # interface
+            bus,
+        )
+
+        if not self.interface:
+            raise DBusException("Could not construct a DBus interface")
+
+    def set_as_presenter_for(self, profile: QWebEngineProfile) -> None:
+        """Sets the profile to use the manager as the presenter."""
+        # PyQtWebEngine unrefs the callback after it's called, for some reason.
+        # So we call setNotificationPresenter again to *increase* its refcount
+        # to prevent it from getting GC'd. Otherwise, random methods start
+        # getting called with the notification as `self`, or segfaults happen,
+        # or other badness.
+        def _present_and_reset(qt_notification: QWebEngineNotification) -> None:
+            profile.setNotificationPresenter(_present_and_reset)
+            self._present(qt_notification)
+
+        profile.setNotificationPresenter(_present_and_reset)
+
+    def _present(self, qt_notification: QWebEngineNotification) -> None:
+        """Shows a notification over DBus.
+
+        This should *not* be directly passed to setNotificationPresenter
+        because of a bug in the PyQtWebEngine bindings.
+        """
+        # notification id 0 means 'assign us the ID'. We can't just pass 0
+        # because it won't get sent as the right type.
+        notification_id = QVariant(0)
+        notification_id.convert(QVariant.UInt)
+
+        actions_list = QDBusArgument([], QMetaType.QStringList)
+
+        qt_notification.show()
+        hints: Dict[str, Any] = {
```

Comment:
You won't be able to use the new-style syntax for variable types as we still support Python 3.5 at the moment.